First tutorial for Atmos (#530)
cloudposse/docs: content/tutorials/atmos-getting-started.md

```yaml
import: []
vars: {}

terraform:
  vars: {}

helmfile:
  vars: {}

components:
  terraform:
    fetch-location:
      vars: {}

    fetch-weather:
      vars:
        api_key: 2a820d40d573758aa714641fc331e897

    output-results:
      vars:
        print_users_weather_enabled: true

  helmfile: {}

workflows:
  deploy-all:
    description: Deploy terraform projects in order
    steps:
      - job: terraform deploy fetch-location
      - job: terraform deploy fetch-weather
      - job: terraform deploy output-results
```

Here we can see our first stack, so let's break this file down to help understand what it is doing:

> 1. We've got a couple empty elements at the top: `import` and `vars`. We'll address these in an upcoming tutorial.

Reviewer: ```suggestion
1. We've got a couple of empty elements at the top: `import` and `vars`. We'll address these in an upcoming tutorial.
```

---
title: "Getting started with Atmos"
description: "Learn what Atmos is and how you can start using it with stacks to simplify your DevOps Automation tasks."
weight: 2
---

## Intro

> `atmos` is a SweetOps built tool to make DevOps and Cloud automation easy. It has direct support for automating Terraform, Helm, Helmfile, and Istio. By natively utilizing [stacks]({{< relref "fundamentals/concepts.md#stacks" >}}), `atmos` enables you to effortlessly manage your Terraform and Helmfile [components]({{< relref "fundamentals/concepts.md#components" >}}) from your local machine or in your CI / CD pipelines.

Reviewer: ```suggestion
`atmos` is part of the SweetOps toolchain and was built to make DevOps and Cloud automation easier across multiple tools. It has direct support for automating Terraform, Helm, Helmfile, and Istio. By natively utilizing [stacks]({{< relref "fundamentals/concepts.md#stacks" >}}), `atmos` enables you to effortlessly manage your Terraform and Helmfile [components]({{< relref "fundamentals/concepts.md#components" >}}) from your local machine or in your CI/CD pipelines.
```

In this tutorial we'll be looking at a simple (albeit contrived) example of automating multiple Terraform components together into a workflow. This will give you some understanding of what `atmos` can do while also giving you some experience with using it at the command line.

## Prerequisites

### Requirements

To accomplish this tutorial, you'll need to have [Git](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git) and [Docker](https://docs.docker.com/get-docker/) installed on your local machine. **That's all**.

### Understanding

> Prior to starting this tutorial, you should be sure you understand [our various concepts and terminology]({{< relref "fundamentals/concepts.md" >}}) and have gone through [Getting started with Geodesic]({{< relref "tutorials/geodesic-getting-started.md" >}}) as we'll be using Geodesic as our means to run `atmos`.

Reviewer: ```suggestion
Prior to starting this tutorial, you should be sure you understand [our various concepts and terminology]({{< relref "fundamentals/concepts.md" >}}) and have gone through our [Getting started with Geodesic]({{< relref "tutorials/geodesic-getting-started.md" >}}) tutorial because we'll be using Geodesic as our means to run `atmos`.
```

## Tutorial

### 1. Clone the tutorials repository

As part of this tutorial (and others in our tutorial series), we'll be utilizing [our tutorials repository](https://github.com/cloudposse/tutorials). This repository includes code and relevant materials for you to use alongside this tutorial walk through.

Let's clone it to your local machine:

```bash
git clone git@github.com:cloudposse/tutorials.git
```

> Now that is on your local machine, let's get into the code that walks us through using `atmos`:

Reviewer: ```suggestion
Now that it's on your local machine, let's get into the code that walks us through using `atmos`:
```

```bash
cd tutorials/02-atmos
```

### 2. Build the image for our tutorial

Now that we've got our code located on your local machine, let's look at our tutorial's example directory:

```
.
├── Dockerfile
├── Makefile
├── README.md
├── components/
└── stacks/
```

Here we've got a `Dockerfile` and a few other things going on. Let's pop open that `Dockerfile` and look at it quickly:

```Dockerfile
ARG VERSION=0.141.6
ARG OS=alpine
ARG CLI_NAME=atmos

FROM cloudposse/geodesic:$VERSION-$OS

# Geodesic message of the Day
ENV MOTD_URL="https://geodesic.sh/motd"

ENV DOCKER_IMAGE="cloudposse/atmos"
ENV DOCKER_TAG="latest"

# Geodesic banner
ENV BANNER="Atmos Tutorial"

# Install terraform.
RUN apk add -u terraform-0.14@cloudposse
# Set Terraform 0.14.x as the default `terraform`.
RUN update-alternatives --set terraform /usr/share/terraform/0.14/bin/terraform

# Install Atmos
RUN apk add atmos@cloudposse

COPY components/ /components/
COPY stacks/ /stacks/

WORKDIR /
```

Few important pieces to point out here:

> 1. We're using [Geodesic]({{< relref "reference/tools.md#geodesic" >}}) as our base image. This enables us to provide a consistent, reproducible toolbox to invoke `atmos` from.

Reviewer: ```suggestion
1. We're using [Geodesic]({{< relref "reference/tools.md#geodesic" >}}) as our base image. This enables us to provide a consistent, reproducible toolbox to invoke `atmos` from the command line.
```

> 1. We're installing the Terraform 0.14 and using that as our default `terraform` executable.

Reviewer: ```suggestion
1. We're installing Terraform 0.14 and using that as our default `terraform` executable. Geodesic supports installing multiple concurrent versions of `terraform` using the [`update-alternatives` system](https://manpages.debian.org/stretch/dpkg/update-alternatives.1.en.html). 
```

1. We're installing `atmos` as a simple binary via `apk`. This is because `atmos` is built and distributed as a [Cloud Posse linux package](https://github.com/cloudposse/packages).
1. We're copying our `components/` and `stacks/` folder into the image.

> Overall, a fairly simple and small set of additions on top of standard Geodesic. To get started using this image, first we have to build it. To do so, we could invoke `docker build` manually, but speed things up we've provided a `make` target to simplify that process:

Reviewer: ```suggestion
Overall, a fairly simple and small set of additions on top of standard Geodesic base image. To get started using this image, first we have to build it. To do so, we could invoke `docker build` manually, but speed things up we've provided a `make` target to simplify that process:
```

```bash
# Pull the Cloud Posse build-harness, which provides some additional utilities
# that our `build` target uses.
make init

# Build our local Dockerfile onto out local machines as `cloudposse/atmos:latest`
make build
```

Once our image is built, we're ready to run it!

### 3. Run the `atmos` image

Now that we've built our image, we want to run it as a new geodesic shell so we can work on our example. Let's run the following command to do that:

```bash
docker run -it \
           --rm \
           --volume $HOME:/localhost \
           --volume $PWD/stacks:/stacks \
           --volume $PWD/components:/components \

>            cloudposse/atmos:latest \
>            --login

Reviewer: @Nuru has updated this so we no longer need to pass login. (although it still works).
```suggestion
           cloudposse/atmos:latest
```

Author: Awesome 👍
